### DISCUSSION TITLE
Error / failure handing in BSP requests

### DISCUSSION THREAD

**Discussion by lefou** [build-server-protocol/build-server-protocol#655] [Proposals] [answered]
The current version of BSP has in my eyes some issues with proper error propagation. Since it's bases on LSP and JSON-RPC, it has in general some error communication channel for each request. Yet, almost all BSP requests are bulk requests, which means they request the same kind of result for many build targets at once. The built-in error channel, the return value, which could be a failed future, does not respect that, so you can either return a success for all requests targets or fail the request completely. There is no way to return some successful build targets and some errors.

Experience has shown, that BSP client tend to heavily use bulk requests. It's rather typical, that a single request for all build targets available is send, especially when initializing the build. If there are even slight issues in these requests, the complete BSP import fails, leaving the user with an unusable editor, with almost nothing working except maybe basic editing.

To proof that, here is an example of a failure, reported against Mill:

* https://github.com/com-lihaoyi/mill/issues/3011

In this is in detail, the `scalacOptions` request failed, as it needed to invoke a compile step as part of it's dependencies, which failed. But since there is no other specified/documented way to return such kind of failure, the complete request resulted in a failure.

There are two essential takeaways from this specific ticket:

1. Some BSP servers seems to handle such issue more graceful that others, beside the protocol having no option for that. They either don't hit the same kind of errors, or they are more failure tolerant and somehow sneak some possibly incomplete or incorrect outcomes into the results. I'd like to highlight, that this isn't necessarily a bad thing, as handling failure that way might be reasonable in specific cases. But it would be nice, if such instances of smart workarounds would be part of the specification/documentation, so we can share some error handling patterns.

2. There is a lack of more fine granular error results in the definition of the BSP responses. I'd expect at least the same granularity level as the request has, which is one failure per build target, or a failure combined with a partial result.

As an alternative, we could also consider to always return results and report errors in a streaming way or by server pushed messages, like we do e.g. for compilation. That way results could be much more granular. It would also improve the responsiveness at the cost of more complexity.

In general, it would be nice, if the BSP specification could address the issue of too large requests and could give recommendations about reasonable request sizes.

**Comment by adpi2** [comment, marked as answer]
I totally agree with this flaw in the design of the multi-target requests and here is my suggestion to solve it: https://github.com/build-server-protocol/build-server-protocol/issues/204
